lightning/src/onion_message/messenger.rs

```diff
@@ -243,6 +244,89 @@ impl OnionMessageRecipient {
 	}
 }
 
+pub trait RespondFunction<T: OnionMessageContents>{
+	fn respond(&self, response: T, reply_path: BlindedPath);
+}
+
+impl<F, T> RespondFunction<T> for F
+where
+T: OnionMessageContents,
+F: Fn(T, BlindedPath)
+{
+	fn respond(&self, response: T, reply_path: BlindedPath) {
+		self(response, reply_path)
+	}
+}
+
+/// A struct handling response to an [`OnionMessage`]
+pub struct Responder<R, T>
+where
+R: RespondFunction<T>,
+T: OnionMessageContents
+{
+	messenger_function: R,
+	pub reply_path: BlindedPath,
+	// This phantom Data is used to ensure that we use T in the struct definition
+	// This allow us to ensure at compile time that the received message type and response type will be same
+	_phantom: PhantomData<T>
+}
+
+impl<R, T> Responder<R, T>
+where
+R: RespondFunction<T>,
+T: OnionMessageContents
+{
+    pub fn respond(&self, response: T) {
+		// Utilising the fact that we ensure at compile time that
+		// received message type, and response type will be same
+		self.messenger_function.respond(response, self.reply_path.clone())
+	}
+}
+
+/// A enum to handle received [`OnionMessage`]
+pub enum ReceivedOnionMessage<R, T>
+where
+R: RespondFunction<T>,
+T: OnionMessageContents
+{
+	WithReplyPath {
+		responder: Responder<R, T>,
+		message: T,
+		path_id: Option<[u8; 32]>
+	},
+	WithoutReplyPath {
+		message: T,
+		path_id: Option<[u8; 32]>,
+	},
+}
+
+impl<R, T> ReceivedOnionMessage<R, T>
+where
+R: RespondFunction<T>,
+T: OnionMessageContents
+{
+	fn new(messenger_function: R, message: T, reply_path_option: Option<BlindedPath>, path_id: Option<[u8; 32]> ) -> Self {
```

Comment:
No need to use `option` in the name here, IMO.

Reply:
I get your point!
I still think having a slightly different name could be handy, especially since we're using reply_path: <BlindedPath> in the function.
I've simplified the name a bit to make it less wordy but still meaningful.
Let me know if we want to simplify it further, and I would love to make the change!